File: frontend/src/features/reports/api/reports.ts
import { api } from '../../../lib/api';
import type { ReportPreviewResponse, ReportType } from '../types';

export const getReportPreview = async (
  type: ReportType,
  startDate?: string,
  endDate?: string
): Promise<ReportPreviewResponse> => {
  const params = new URLSearchParams();
  params.append('type', type);
  if (startDate) params.append('start_date', startDate);
  if (endDate) params.append('end_date', endDate);
  
  // api.get returns the generic T directly, and API_BASE_URL doesn't include /api
  const data = await api.get<ReportPreviewResponse>(`/api/reports/preview?${params.toString()}`);
  return data;
};

export const exportReportExcel = async (
  type: ReportType,
  startDate?: string,
  endDate?: string
): Promise<void> => {
  const params = new URLSearchParams();
  params.append('type', type);
  if (startDate) params.append('start_date', startDate);
  if (endDate) params.append('end_date', endDate);
  
  // Use the built-in api.download method from lib/api.ts
  await api.download(`/api/reports/export?${params.toString()}`, `reporte_${type}.xlsx`);
};

File: backend/app/api/reports.py
"""
Reports endpoints.

Phase 11 will implement:
- GET    /api/reports/preview      — Generate report data (with date/service filters)
- GET    /api/reports/export       — Export report as XLSX
"""

from fastapi import APIRouter, Depends, Query
from fastapi.responses import StreamingResponse

from app.core.dependencies import DBDep, require_staff
from app.schemas.report import ReportPreviewResponse
from app.services.report_service import get_report_data, export_report_excel

# Apply require_staff to all endpoints in this router
router = APIRouter(dependencies=[Depends(require_staff)])

@router.get("/preview", response_model=ReportPreviewResponse, summary="Preview report")
async def preview_report(
    db: DBDep,
    type: str = Query(..., description="Report type: loans, cubicles, visitors, events, activities"),
    start_date: str | None = Query(None),
    end_date: str | None = Query(None),
    limit: int = Query(100)
):
    """Generate preview report data."""
    columns, rows = get_report_data(db, report_type=type, start_date=start_date, end_date=end_date, limit=limit)
    return ReportPreviewResponse(columns=columns, rows=rows)

@router.get("/export", summary="Export report as Excel")
async def export_report(
    db: DBDep,
    type: str = Query(..., description="Report type"),
    start_date: str | None = Query(None),
    end_date: str | None = Query(None)
):
    """Generate and download XLSX file."""
    buffer = export_report_excel(db, report_type=type, start_date=start_date, end_date=end_date)
    
    return StreamingResponse(
        buffer,
        media_type="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        headers={"Content-Disposition": f"attachment; filename=reporte_{type}.xlsx"}
    )
